test: Remove mock so that future tests will not be affected by the mock

danieljbruce opened the pull request:
This issue was discovered when trying to wrap `test/index` in async. Tests were failing because a mock in one test was being used again in future tests and caused those tests to fail in the second async loop. The change involves resetting the mock to its original setup so that future tests do not use the mock that was set up in that one particular test.

danieljbruce committed:
Ensure the mock isn’t permanent for other cases

Reassign this mock back to what it was because future tests will rely on this being the proper value.

danieljbruce committed:
Add async to verify the change works

Async will run all the tests twice failing if the prior change was not implemented.

danieljbruce committed:
Revert "Add async to verify the change works"

This reverts commit 57855a32bc9ca0ac19cd310d593869801a9a7fd3.

danieljbruce committed:
Adjust the comment slightly

Adjust the comment slightly to address the fact that we are changing the mocked function back to the way it was.

danieljbruce committed:
Refactor the mock out into a constant

The mock should be a constant instead of repeating a code fragment that must be the same.

kolea2 commented on test/index.ts:
```diff
@@ -1774,6 +1776,8 @@ describe('Datastore', () => {
           excludeFromIndexes
         );
+        // Reassign this to what it was so that other tests don't fail because of the mock.
+        fakeEntity.entityToEntityProto = ENTITY_TO_ENTITY_PROTO;
```
maybe surround this test with a try/finally block to ensure this resets even if the test fails?

danieljbruce replied:
Done

danieljbruce replied:
Reverted. Others had comments so may chime in a bit later.

vicb replied:
I don't think `try`/`finally` would be a good fit here because it's sync and the test is async (that is why might reset the prop before it is actually used).

I am looked at this super closely but there are several things to explore to reset the prop IMO:
- use an `async` test fn if the test framework and the code flow support it,
- use `afterTest` + global objet to restore explicitly saved props (that would look a little hacky)

danieljbruce replied:
By the last comment do you mean wrapping this test in a describe block and then adding an after hook in the describe block?

vicb replied:
Actually I don't think my first option would work.

Maybe something to explore:

```js
const fakeEntityInit: any = {
 // previous fakeEntity
};

const fakeEntity: any = {};

// ...

  beforeEach(() => {
    // init fakeEntity
    fakeEntity[KEY_SYMBOL] = fakeEntityInit[KEY_SYMBOL];
    for (const [key, value] of Object.entries(fakeEntityInit)) {
      fakeEntity[key] = value;
    }

    // existing code   
    createInsecureOverride = null;
    googleAuthOverride = null;
    datastore = new Datastore({
      projectId: PROJECT_ID,
      namespace: NAMESPACE,
    });
  });
```

This would ensure the `fakeEntity` is init from a clean value before each test.

danieljbruce replied:
This would work. I guess we don't need to care too much about the extra run time this would add since these are tests.

danieljbruce replied:
I adopted the same idea, but did it a slightly different way. We do not want to have to iterate through each property like this. We can just use `Object.assign` and that will do the trick.

danieljbruce committed:
Surround with try/finally

We can use try/finally so that even if the test fails the mock gets reset.

danieljbruce committed:
Revert "Surround with try/finally"

This reverts commit 73bea28cd0d9998e1b670fe09f5ea896c39d72b1.

danieljbruce committed:
Have a fake entity init state

Have a fake entity init state and then before each test initialize the fakeEntity from the fakeEntityInit state.

danieljbruce committed:
Remove resetting the mock in individual tests

Now that we initialize the fake entity at the beginning of every single test, we do not need to reset the mock in that one test anymore.

danieljbruce committed:
Eliminate unused variable

Variable is not used anymore so should be eliminated from the code change.

danieljbruce committed:
Merge branch 'main' into test-async-prep
